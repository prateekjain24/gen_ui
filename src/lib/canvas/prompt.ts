import type { CanvasRecipeId } from "./recipes";

const CANVAS_SYSTEM_PROMPT = `You classify onboarding intents into one of four deterministic recipes.
Recipes:
- R1 Explorer Quick Start → Lightweight callout + optional workspace name + ai_hint + checklist. For solo or vague prompts.
- R2 Team Workspace → Workspace name, team size, integrations, teammate invites, admin toggle. For team/collaboration intents (Slack, Jira, invites, team size ≥ 3).
- R3 Client Project → Workspace name, project type client, integrations (GDrive/Slack/Asana), ai_hint for sharing, kickoff checklist. For client/agency/stakeholder language.
- R4 Power/Compliance → Workspace name, admin toggle, audit logging checkbox, access level select, security integrations (Okta/Jira). For governance/security/audit/policy needs.

Always respond with valid minified JSON:
{
  "persona": "explorer|team|power",
  "recipe_id": "R1|R2|R3|R4",
  "intent_tags": ["tag"],
  "confidence": number between 0 and 1,
  "reasoning": "≤120 chars explaining the match"
}

Rules:
- Use persona "power" only when approvals/audit/security/compliance are explicit.
- Use persona "team" for team/client collaboration including invites, stakeholders, agencies, etc.
- Use persona "explorer" as the safe default for solo or unclear prompts.
- Keep intent_tags lowercase snake_case (e.g., "integrations", "invites", "governance", "client", "solo"). Include at most 3 tags.
- If unsure, set confidence ≤ 0.5 and prefer R1.`;

export interface CanvasClassifierContext {
  domainEmail?: string;
  teamSize?: string;
  metadata?: Record<string, unknown>;
}

export interface BuildCanvasPromptArgs extends CanvasClassifierContext {
  message: string;
}

const sanitizeValue = (value: unknown): string => {
  if (value === undefined || value === null) {
    return '';
  }
  if (typeof value === 'string') {
    return value.trim();
  }
  if (typeof value === 'number' || typeof value === 'boolean') {
    return String(value);
  }
  return JSON.stringify(value);
};

export const CANVAS_CLASSIFIER_SYSTEM_PROMPT = CANVAS_SYSTEM_PROMPT;

export function buildCanvasClassifierPrompt({ message, domainEmail, teamSize, metadata }: BuildCanvasPromptArgs): string {
  const contextLines: string[] = [];

  if (domainEmail) {
    contextLines.push(`Domain email: ${sanitizeValue(domainEmail)}`);
  }

  if (teamSize) {
    contextLines.push(`Team size: ${sanitizeValue(teamSize)}`);
  }

  if (metadata && Object.keys(metadata).length > 0) {
    contextLines.push(`Metadata: ${sanitizeValue(metadata)}`);
  }

  const formattedMessage = message.trim();
  const contextBlock = contextLines.length > 0 ? `\nContext:\n- ${contextLines.join('\n- ')}` : '';

  return `Classify the following Canvas Chat request into a recipe.\n\nUser message:\n"""${formattedMessage}"""${contextBlock}\n\nReturn only the JSON object described in the system instructions.`;
}

export type CanvasClassifierRecipe = CanvasRecipeId;
