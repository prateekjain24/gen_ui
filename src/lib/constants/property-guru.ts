/**
 * PropertyGuru-specific constant lookups shared across extractor, UI, and telemetry.
 */

import type {
  PropertyGuruLifestyleCue,
  PropertyGuruMoveInHorizon,
  PropertyGuruPropertyType,
  PropertyGuruTonePreset,
} from '@/lib/types/property-guru';

export const PROPERTY_GURU_PROPERTY_TYPES: readonly PropertyGuruPropertyType[] = [
  'condo',
  'executive_condo',
  'hdb',
  'landed',
  'serviced_apartment',
  'co_living',
  'other',
] as const;

export const PROPERTY_GURU_LIFESTYLE_CUES: readonly PropertyGuruLifestyleCue[] = [
  'near_mrt',
  'near_schools',
  'childcare_ready',
  'family_friendly',
  'pet_friendly',
  'wheelchair_access',
  'healthcare_access',
  'green_spaces',
  'co_working',
  'serviced_amenities',
  'rental_hotspot',
  'move_in_ready',
  'renovation_friendly',
  'dual_city',
  'community_events',
] as const;

export const PROPERTY_GURU_TONE_PRESETS: readonly PropertyGuruTonePreset[] = [
  'reassuring',
  'data_driven',
  'concierge',
] as const;

export const PROPERTY_GURU_MOVE_IN_HORIZONS: readonly PropertyGuruMoveInHorizon[] = [
  'immediate',
  'three_months',
  'six_months',
  'year_plus',
  'flexible',
  'unspecified',
] as const;
