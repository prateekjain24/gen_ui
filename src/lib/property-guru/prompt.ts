import { z } from 'zod';

import type { PropertyGuruSignals } from '@/lib/types/property-guru';

export const PROPERTY_GURU_PLAN_SYSTEM_PROMPT = `You are a PropertyGuru onboarding planner helping home seekers turn a conversational brief into an actionable discovery plan.

Always respond with JSON that matches the provided schema. Your plan must feel like a concierge advisor who understands the seeker. Reflect the user's tone preference when available.

### Plan structure
1. intent_summary — 1-2 sentences affirming the seeker's goal.
2. essentials — key filters to tune first (location radius, price band, property type/bedrooms). Present as labelled items with short guidance and recommended values.
3. lifestyle_filters — highlight lifestyle toggles worth adjusting (amenities, accessibility, commute, family cues). Explain why each matters.
4. plan_actions — primary and optional secondary actions to continue in PropertyGuru (e.g., view matches, start mortgage check, schedule tour). Include CTA labels and contextual helper copy.
5. micro_copy — reassurance + follow-up language (listing freshness, weekly digest, concierge support).

### Guardrails
- Never invent impossible data; stay within the signals provided.
- Use Singapore context (currency SGD, MRT references, districts).
- Tone must be one of: reassuring, data-driven, concierge.
- Keep each string ≤ 200 characters.
- Prefer bullet-like phrasing separated by semicolons when listing multiple benefits.
- If information is missing, note sensible defaults (e.g., "Budget not specified — starting with SGD 1.0M").

Return JSON only.`;

export const propertyGuruPlanSchema = z.object({
  intent_summary: z.string().min(1),
  essentials: z.object({
    title: z.string().min(1),
    items: z
      .array(
        z.object({
          label: z.string().min(1),
          value: z.string().min(1),
          helper: z.string().min(1),
        })
      )
      .min(1),
  }),
  lifestyle_filters: z.object({
    title: z.string().min(1),
    highlights: z.array(z.string().min(1)).min(1),
    helper: z.string().min(1),
  }),
  plan_actions: z.object({
    primary: z.object({
      label: z.string().min(1),
      description: z.string().min(1),
    }),
    secondary: z
      .array(
        z.object({
          label: z.string().min(1),
          description: z.string().min(1),
        })
      )
      .default([]),
  }),
  micro_copy: z.object({
    reassurance: z.string().min(1),
    follow_up: z.string().min(1),
  }),
});

export type PropertyGuruPlanTemplate = z.infer<typeof propertyGuruPlanSchema>;

export const buildPropertyGuruPlanPrompt = ({
  prompt,
  signals,
}: {
  prompt: string;
  signals: PropertyGuruSignals;
}): string => {
  const summary = {
    prompt,
    signals,
  };

  return `Context: ${JSON.stringify(summary, null, 2)}\n\nReturn a JSON object that matches the PropertyGuru schema.`;
};

export const createDefaultPropertyGuruPlan = (): PropertyGuruPlanTemplate => ({
  intent_summary: 'Let\'s refine your home search together.',
  essentials: {
    title: 'Tune the essentials',
    items: [
      {
        label: 'Location radius',
        value: '5km around your preferred neighbourhood',
        helper: 'Adjust the slider to widen or tighten nearby listings.',
      },
      {
        label: 'Price band',
        value: 'SGD 1.0M - 1.2M',
        helper: 'Set a range that keeps monthly payments comfortable.',
      },
    ],
  },
  lifestyle_filters: {
    title: 'Lifestyle filters',
    highlights: ['Commute to MRT', 'Nearby schools', 'Pet-friendly options'],
    helper: 'Toggle the priorities that matter; we will reshuffle matches instantly.',
  },
  plan_actions: {
    primary: {
      label: 'Preview curated listings',
      description: 'Opens a shortlist with filters applied.',
    },
    secondary: [
      {
        label: 'Start mortgage check',
        description: 'See monthly repayments and financing options.',
      },
    ],
  },
  micro_copy: {
    reassurance: 'Listings refresh daily so you don\'t miss new launches.',
    follow_up: 'We will email a weekly digest and concierge tips.',
  },
});
