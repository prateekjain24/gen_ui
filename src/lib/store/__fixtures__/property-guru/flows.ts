import type { PropertyGuruPlanTemplate } from '@/lib/property-guru/prompt';
import type { PropertyGuruLifestyleCue, PropertyGuruTonePreset } from '@/lib/types/property-guru';

interface PropertyGuruFlowExpectations {
  location: string;
  propertyType?: string;
  bedrooms?: number;
  lifestyleIncludes?: PropertyGuruLifestyleCue[];
  tone?: PropertyGuruTonePreset;
  nextSteps: string[];
  copyTone: PropertyGuruTonePreset;
  primaryCta: string;
  secondaryCtas?: string[];
  defaultsApplied?: string[];
}

export interface PropertyGuruFlowFixture {
  id: string;
  prompt: string;
  plan: PropertyGuruPlanTemplate;
  expectations: PropertyGuruFlowExpectations;
}

export const PROPERTY_GURU_FLOW_FIXTURES: PropertyGuruFlowFixture[] = [
  {
    id: 'twins_near_tampines',
    prompt:
      "Expecting twins in 4 months, want a 3BR condo near Tampines MRT, budget around 1.4M but I'll stretch for turnkey. Need good preschool options and low-renovation.",
    plan: {
      intent_summary: 'Focused search: 3BR turnkey condo near Tampines MRT within ~SGD 1.4M.',
      essentials: {
        title: 'Tune the essentials first',
        items: [
          {
            label: 'Location radius',
            value: '1.5km around Tampines MRT',
            helper: 'Keeps you within walking distance and near childcare hubs.',
          },
          {
            label: 'Price band',
            value: 'SGD 1.2M – 1.4M (stretch +5% for turnkey units)',
            helper: 'Set alerts at SGD 1.45M so turnkey options surface quickly.',
          },
          {
            label: 'Bedrooms & layout',
            value: '3 bedrooms, turnkey finish',
            helper: 'Filter for renovated family layouts with dual nursery potential.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Lifestyle filters to toggle',
        highlights: [
          'Childcare-ready developments',
          'Family-friendly facilities',
          'Low-renovation inventory',
        ],
        helper: 'Prioritise projects near preschools with playgrounds and service corridors.',
      },
      plan_actions: {
        primary: {
          label: 'Preview curated Tampines listings',
          description: 'Opens a shortlist with MRT proximity and school filters applied.',
        },
        secondary: [
          {
            label: 'Book PropertyGuru Finance consult',
            description: 'Align financing buffers for turnkey units and twins arrival timeline.',
          },
          {
            label: 'Schedule neighbourhood tour',
            description: 'Compare childcare options and transport during a guided visit.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'Listings refresh daily; concierge flags new turnkey options instantly.',
        follow_up: 'Saved searches send weekly MRT-adjacent family picks plus preschool intel.',
      },
    },
    expectations: {
      location: 'Tampines',
      propertyType: 'condo',
      bedrooms: 3,
      lifestyleIncludes: ['near_mrt', 'childcare_ready', 'family_friendly', 'move_in_ready'],
      nextSteps: ['viewListings', 'mortgagePreCheck', 'scheduleTour'],
      copyTone: 'reassuring',
      primaryCta: 'Preview curated Tampines listings',
      secondaryCtas: ['Book PropertyGuru Finance consult', 'Schedule neighbourhood tour'],
    },
  },
  {
    id: 'split_city_worker',
    prompt:
      'Splitting time between Singapore and KL, need a 1BR serviced apartment along the Thomson-East Coast TEL line, want co-working nearby and flexible lease; budget SGD 3.2K monthly.',
    plan: {
      intent_summary: 'Balance your split-city lifestyle with serviced TEL options.',
      essentials: {
        title: 'Lock these in first',
        items: [
          {
            label: 'TEL anchor',
            value: 'Within 1km of Thomson-East Coast Line stations',
            helper: 'Stay near direct airport links for quick KL runs.',
          },
          {
            label: 'Monthly lease budget',
            value: 'SGD 3,200 serviced apartment',
            helper: 'Keep rent aligned with flexible stay expectations.',
          },
          {
            label: 'Layout',
            value: '1 bedroom serviced apartment',
            helper: 'Ensure concierge, housekeeping, and on-demand amenities.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Remote work essentials',
        highlights: ['Co-working access', 'Serviced amenities', 'Flexible lease terms'],
        helper: 'Toggle extras when you need concierge support or hot desk passes.',
      },
      plan_actions: {
        primary: {
          label: 'View serviced apartments near TEL',
          description: 'See TEL-adjacent units with concierge and co-working perks.',
        },
        secondary: [
          {
            label: 'Arrange remote video tour',
            description: 'Line up virtual walkthroughs while you are in KL.',
          },
          {
            label: 'Book PropertyGuru concierge consult',
            description: 'Get help balancing flexible lease terms across cities.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'We will flag concierge-ready TEL units as soon as they refresh.',
        follow_up: 'Expect mid-week digests and concierge nudges when new serviced stock appears.',
      },
    },
    expectations: {
      location: 'Thomson-east Coast Line',
      propertyType: 'serviced_apartment',
      bedrooms: 1,
      lifestyleIncludes: ['dual_city', 'co_working', 'serviced_amenities'],
      nextSteps: ['viewListings', 'scheduleTour', 'expertConsult'],
      copyTone: 'reassuring',
      primaryCta: 'View serviced apartments near TEL',
      secondaryCtas: ['Arrange remote video tour', 'Book PropertyGuru concierge consult'],
      defaultsApplied: ['filters.districts'],
    },
  },
  {
    id: 'yield_investor',
    prompt:
      'Looking for a 2BR resale condo in Queenstown under 1.3M, strong rental demand, okay with minor reno, want data on yield and past transactions.',
    plan: {
      intent_summary: 'Spot 2BR yield winners in Queenstown under SGD 1.3M.',
      essentials: {
        title: 'Dial in the fundamentals',
        items: [
          {
            label: 'District focus',
            value: 'District 3 – Queenstown resale stock',
            helper: 'Locks in proven rental demand around MRT and business parks.',
          },
          {
            label: 'Price ceiling',
            value: 'SGD 1.25M – 1.3M',
            helper: 'Keep listings inside your yield target while leaving room for fees.',
          },
          {
            label: 'Layout',
            value: '2 bedroom resale condo',
            helper: 'Choose efficient floorplans tenants favour.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Investor toggles',
        highlights: ['Rental hotspot indicators', 'Reno-friendly layouts', 'Yield comparables'],
        helper: 'Switch on reports and recent transactions while you shortlist.',
      },
      plan_actions: {
        primary: {
          label: 'Review high-yield Queenstown listings',
          description: 'Shortlist data-backed units with tenant demand signals.',
        },
        secondary: [
          {
            label: 'Download rental transaction digest',
            description: 'Export recent transactions and yield benchmarks.',
          },
          {
            label: 'Book investor strategy consult',
            description: 'Walk through financing and renovation math with specialists.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'We surface units where yield forecasts and tenant demand stay strong.',
        follow_up: 'Receive new comparables and rent trend alerts every week.',
      },
    },
    expectations: {
      location: 'Queenstown',
      propertyType: 'condo',
      bedrooms: 2,
      lifestyleIncludes: ['rental_hotspot'],
      tone: 'data_driven',
      nextSteps: ['viewListings', 'expertConsult'],
      copyTone: 'data_driven',
      primaryCta: 'Review high-yield Queenstown listings',
      secondaryCtas: ['Download rental transaction digest', 'Book investor strategy consult'],
    },
  },
  {
    id: 'multigenerational_accessibility',
    prompt:
      'Retiring parents moving in, need 4BR near Bishan, lift access, wheelchair friendly, near hospitals and parks; budget 2.3M; we can wait 6-9 months.',
    plan: {
      intent_summary: 'Plan a Bishan move-in that keeps the whole family comfortable.',
      essentials: {
        title: 'Lock these in early',
        items: [
          {
            label: 'Location radius',
            value: '3km around Bishan with hospital access',
            helper: 'Keeps you minutes from healthcare and park connectors.',
          },
          {
            label: 'Layout & access',
            value: '4 bedroom homes with lift access',
            helper: 'Filter for barrier-free entries and elder-friendly blocks.',
          },
          {
            label: 'Budget',
            value: 'Up to SGD 2.3M with buffer for improvements',
            helper: 'Leaves room for accessibility retrofits if needed.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Comfort toggles',
        highlights: ['Wheelchair-friendly routes', 'Hospitals within 10 minutes', 'Park connectors & greenery'],
        helper: 'Keep wellness cues on so the plan balances care and downtime.',
      },
      plan_actions: {
        primary: {
          label: 'View accessibility-ready Bishan homes',
          description: 'Opens listings vetted for lift access and elder-friendly layouts.',
        },
        secondary: [
          {
            label: 'Schedule hospital-nearby tour',
            description: 'Plot a visit that checks commute times to clinics.',
          },
          {
            label: 'Book accessibility renovation consult',
            description: 'Line up pros for grab bars or layout tweaks before move-in.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'We emphasise accessibility and calm routes so parents settle in comfortably.',
        follow_up: 'Expect fortnightly updates while you wait for the right listing.',
      },
    },
    expectations: {
      location: 'Bishan',
      bedrooms: 4,
      lifestyleIncludes: ['wheelchair_access', 'healthcare_access', 'green_spaces'],
      nextSteps: ['viewListings', 'scheduleTour', 'expertConsult'],
      copyTone: 'reassuring',
      primaryCta: 'View accessibility-ready Bishan homes',
      secondaryCtas: ['Schedule hospital-nearby tour', 'Book accessibility renovation consult'],
    },
  },
  {
    id: 'first_time_ec',
    prompt:
      'First time buying, looking at executive condos (EC) in Punggol around 1M, not pre-approved, two kids in primary school, want to stay near primary schools, want grants guidance and weekly listing digests.',
    plan: {
      intent_summary: 'Guide your first EC purchase in Punggol with grants in mind.',
      essentials: {
        title: 'Start with these',
        items: [
          {
            label: 'Location focus',
            value: 'Punggol ECs near schools',
            helper: 'Stay close to primary schools and community hubs.',
          },
          {
            label: 'Budget band',
            value: 'Around SGD 1.0M with grant support',
            helper: 'Keep monthly repayments friendly while grants process.',
          },
          {
            label: 'Financing prep',
            value: 'Line up HDB and PropertyGuru Finance guidance',
            helper: 'Know your grants and approval path before locking a unit.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Family priorities',
        highlights: ['Nearby primary schools', 'Family amenities', 'Weekly digest ready'],
        helper: 'Keep these toggled so the plan nudges you with kid-friendly picks.',
      },
      plan_actions: {
        primary: {
          label: 'Save your EC shortlist',
          description: 'Create a shortlist that refreshes as new launches appear.',
        },
        secondary: [
          {
            label: 'Start HDB grant checklist',
            description: 'Work through grant eligibility with guided steps.',
          },
          {
            label: 'Book PropertyGuru mortgage chat',
            description: 'Get pre-approval coaching before you commit.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'We will nudge you on grants and eligibility as you progress.',
        follow_up: 'Weekly digests keep new EC options top of mind.',
      },
    },
    expectations: {
      location: 'Punggol',
      propertyType: 'executive_condo',
      lifestyleIncludes: ['near_schools', 'family_friendly'],
      nextSteps: ['savedSearch', 'startHdbGrantChecklist'],
      copyTone: 'reassuring',
      primaryCta: 'Save your EC shortlist',
      secondaryCtas: ['Start HDB grant checklist', 'Book PropertyGuru mortgage chat'],
    },
  },
];
