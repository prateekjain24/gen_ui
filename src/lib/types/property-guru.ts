/**
 * PropertyGuru seeker signal definitions
 *
 * Shared contract describing the typed signals the PropertyGuru preset
 * expects. This keeps extractors, prompt engineering, UI, and telemetry
 * aligned so new work avoids ad hoc string keys.
 */

export type PropertyGuruTonePreset = 'reassuring' | 'data_driven' | 'concierge';

export type PropertyGuruPropertyType =
  | 'condo'
  | 'executive_condo'
  | 'hdb'
  | 'landed'
  | 'serviced_apartment'
  | 'co_living'
  | 'other';

export type PropertyGuruLifestyleCue =
  | 'near_mrt'
  | 'near_schools'
  | 'childcare_ready'
  | 'family_friendly'
  | 'pet_friendly'
  | 'wheelchair_access'
  | 'healthcare_access'
  | 'green_spaces'
  | 'co_working'
  | 'serviced_amenities'
  | 'rental_hotspot'
  | 'move_in_ready'
  | 'renovation_friendly'
  | 'dual_city'
  | 'community_events';

export type PropertyGuruFinanceReadiness =
  | 'pre_approved'
  | 'exploring_options'
  | 'needs_guidance'
  | 'cash_purchase'
  | 'unknown';

export type PropertyGuruMoveInHorizon =
  | 'immediate'
  | 'three_months'
  | 'six_months'
  | 'year_plus'
  | 'flexible'
  | 'unspecified';

export type PropertyGuruBedroomCount = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export interface PropertyGuruLocationFocus {
  /** Primary area, neighbourhood, or planning region name */
  primaryArea: string;
  /** Optional secondary areas to consider */
  secondaryAreas?: string[];
  /** Specific landmarks or transport anchors (e.g. MRT lines) */
  anchorPoints?: string[];
  /** Preferred search radius in kilometres */
  radiusKm?: number;
  /** Optional PropertyGuru district identifiers */
  districtIds?: number[];
}

export interface PropertyGuruPriceBand {
  /** Minimum budget in currency units */
  min?: number;
  /** Maximum comfortable budget */
  max?: number;
  /** Currency code, defaulting to SGD */
  currency: 'SGD';
  /** Whether seeker is open to stretching beyond max (~10%) */
  stretchOk?: boolean;
}

export interface PropertyGuruSignals {
  location: PropertyGuruLocationFocus;
  price: PropertyGuruPriceBand;
  propertyType: PropertyGuruPropertyType;
  bedrooms: PropertyGuruBedroomCount;
  moveInHorizon: PropertyGuruMoveInHorizon;
  lifestyle: PropertyGuruLifestyleCue[];
  financeReadiness: PropertyGuruFinanceReadiness;
  tonePreference: PropertyGuruTonePreset;
}

export type PropertyGuruSignalsPartial = Partial<PropertyGuruSignals>;
