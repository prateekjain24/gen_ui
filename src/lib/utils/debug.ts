/**
 * Debug utilities for development-only logging
 *
 * Provides conditional logging that only outputs in development mode,
 * preventing console statements from appearing in production builds.
 */

/**
 * Checks if the current environment is development
 */
const isDevelopment = process.env.NODE_ENV === 'development';

/**
 * Debug logger that only outputs in development mode
 *
 * @param message - The primary message to log
 * @param args - Additional arguments to log
 *
 * @example
 * debug('Session created', sessionId);
 * debug('[SessionStore]', 'Cleanup completed', { count: 5 });
 */
export function debug(message: string, ...args: unknown[]): void {
  if (isDevelopment) {
    // eslint-disable-next-line no-console
    console.log(message, ...args);
  }
}

/**
 * Debug warning that only outputs in development mode
 *
 * @param message - The warning message
 * @param args - Additional arguments
 */
export function debugWarn(message: string, ...args: unknown[]): void {
  if (isDevelopment) {
    console.warn(message, ...args);
  }
}

/**
 * Debug error that always outputs (even in production)
 *
 * @param message - The error message
 * @param args - Additional arguments
 */
export function debugError(message: string, ...args: unknown[]): void {
  console.error(message, ...args);
}

/**
 * Creates a namespaced debug logger
 *
 * @param namespace - The namespace for the logger
 * @returns A debug function for that namespace
 *
 * @example
 * const log = createDebugger('SessionStore');
 * log('Created session', sessionId);
 * // Output: [SessionStore] Created session abc123
 */
export function createDebugger(namespace: string) {
  return (message: string, ...args: unknown[]) => {
    debug(`[${namespace}] ${message}`, ...args);
  };
}

/**
 * Conditional debug logger based on a flag
 *
 * @param condition - Whether to log
 * @param message - The message to log
 * @param args - Additional arguments
 */
export function debugIf(condition: boolean, message: string, ...args: unknown[]): void {
  if (condition) {
    debug(message, ...args);
  }
}

/**
 * Performance timer for development debugging
 *
 * @param label - The label for the timer
 * @returns A function to end the timer
 *
 * @example
 * const endTimer = debugTimer('API Call');
 * await fetchData();
 * endTimer(); // Logs: [Timer] API Call: 234ms
 */
export function debugTimer(label: string): () => void {
  if (!isDevelopment) {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    return () => {}; // No-op in production
  }

  const start = performance.now();
  return () => {
    const duration = performance.now() - start;
    debug(`[Timer] ${label}: ${duration.toFixed(2)}ms`);
  };
}