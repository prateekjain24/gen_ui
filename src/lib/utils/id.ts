/**
 * ID generation utilities for the Generative UI POC
 *
 * Provides consistent UUID generation and validation across the application.
 * Uses UUID v4 for random, collision-resistant identifiers.
 */

import { v4 as uuidv4 } from 'uuid';

/**
 * Generates a new UUID v4
 *
 * @returns A random UUID v4 string
 * @example
 * const sessionId = generateId();
 * // Returns: "550e8400-e29b-41d4-a716-446655440000"
 */
export function generateId(): string {
  return uuidv4();
}

/**
 * Generates a prefixed ID for specific entity types
 *
 * @param prefix - The prefix to add (e.g., "session", "event")
 * @returns A prefixed UUID string
 * @example
 * const sessionId = generatePrefixedId('session');
 * // Returns: "session_550e8400-e29b-41d4-a716-446655440000"
 */
export function generatePrefixedId(prefix: string): string {
  return `${prefix}_${uuidv4()}`;
}

/**
 * Validates if a string is a valid UUID v4
 *
 * @param id - The string to validate
 * @returns True if the string is a valid UUID v4
 */
export function isValidUUID(id: string): boolean {
  const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
  return uuidRegex.test(id);
}

/**
 * Validates if a string is a valid prefixed UUID
 *
 * @param id - The string to validate
 * @param prefix - The expected prefix
 * @returns True if the string has the correct prefix and valid UUID
 */
export function isValidPrefixedId(id: string, prefix: string): boolean {
  if (!id.startsWith(`${prefix}_`)) {
    return false;
  }
  const uuid = id.slice(prefix.length + 1);
  return isValidUUID(uuid);
}

/**
 * Generates a short ID suitable for user-facing purposes
 *
 * @param length - The desired length (default: 8)
 * @returns A short random string
 * @example
 * const shortId = generateShortId();
 * // Returns: "a3k9x2p1"
 */
export function generateShortId(length: number = 8): string {
  const chars = '0123456789abcdefghijklmnopqrstuvwxyz';
  let result = '';
  for (let i = 0; i < length; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

/**
 * Extracts the UUID from a prefixed ID
 *
 * @param prefixedId - The prefixed ID string
 * @returns The UUID portion or null if invalid
 */
export function extractUUID(prefixedId: string): string | null {
  const match = prefixedId.match(/[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}/i);
  return match ? match[0] : null;
}