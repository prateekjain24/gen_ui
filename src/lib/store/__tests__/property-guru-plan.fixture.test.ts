import { buildPropertyGuruPlanPrompt, propertyGuruPlanSchema } from '@/lib/property-guru/prompt';
import { extractPropertyGuruSignals } from '@/lib/utils/property-guru-signals';

const TAMPINES_PROMPT =
  'Expecting twins in 4 months, want a 3BR condo near Tampines MRT, budget around 1.4M but I\'ll stretch for turnkey. Need good preschool options and low-reno.';

describe('PropertyGuru plan preset fixtures', () => {
  it('produces prompt context containing extracted signals', () => {
    const { signals } = extractPropertyGuruSignals(TAMPINES_PROMPT);
    const planPrompt = buildPropertyGuruPlanPrompt({ prompt: TAMPINES_PROMPT, signals });

    expect(planPrompt).toMatchSnapshot();
  });

  it('validates golden plan structure for the Tampines MRT twins scenario', () => {
    const plan = propertyGuruPlanSchema.parse({
      intent_summary: 'Focused search: 3BR turnkey condo near Tampines MRT within ~SGD 1.4M.',
      essentials: {
        title: 'Tune the essentials first',
        items: [
          {
            label: 'Location radius',
            value: '1.5km around Tampines MRT',
            helper: 'Keeps you within walking distance and near childcare hubs.',
          },
          {
            label: 'Price band',
            value: 'SGD 1.2M â€“ 1.4M (stretch +5% for turnkey units)',
            helper: 'Set alerts at SGD 1.45M so turnkey options surface quickly.',
          },
          {
            label: 'Bedrooms & layout',
            value: '3 bedrooms, turnkey finish',
            helper: 'Filter for renovated family layouts with dual nursery potential.',
          },
        ],
      },
      lifestyle_filters: {
        title: 'Lifestyle filters to toggle',
        highlights: [
          'Childcare-ready developments',
          'Family-friendly facilities',
          'Low-renovation inventory',
        ],
        helper: 'Prioritise projects near preschools with playgrounds and service corridors.',
      },
      plan_actions: {
        primary: {
          label: 'Preview curated Tampines listings',
          description: 'Opens a shortlist with MRT proximity and school filters applied.',
        },
        secondary: [
          {
            label: 'Book PropertyGuru Finance consult',
            description: 'Align financing buffers for turnkey units and twins arrival timeline.',
          },
          {
            label: 'Schedule neighbourhood tour',
            description: 'Compare childcare options and transport during a guided visit.',
          },
        ],
      },
      micro_copy: {
        reassurance: 'Listings refresh daily; concierge flags new turnkey options instantly.',
        follow_up: 'Saved searches send weekly MRT-adjacent family picks plus preschool intel.',
      },
    });

    expect(plan).toMatchSnapshot();
  });
});
