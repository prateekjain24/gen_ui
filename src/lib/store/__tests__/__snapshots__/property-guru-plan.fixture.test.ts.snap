// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PropertyGuru plan preset fixtures produces prompt context containing extracted signals 1`] = `
"Context: {
  "prompt": "Expecting twins in 4 months, want a 3BR condo near Tampines MRT, budget around 1.4M but I'll stretch for turnkey. Need good preschool options and low-reno.",
  "signals": {
    "location": {
      "primaryArea": "Tampines",
      "radiusKm": 1.5,
      "anchorPoints": [
        "Tampines MRT"
      ],
      "districtIds": [
        18
      ]
    },
    "price": {
      "currency": "SGD",
      "max": 1400000,
      "min": 1400000,
      "stretchOk": true
    },
    "propertyType": "condo",
    "bedrooms": 3,
    "moveInHorizon": "six_months",
    "lifestyle": [
      "near_mrt",
      "near_schools",
      "childcare_ready",
      "family_friendly",
      "move_in_ready"
    ],
    "financeReadiness": "unknown",
    "tonePreference": "reassuring"
  }
}

Return a JSON object that matches the PropertyGuru schema."
`;

exports[`PropertyGuru plan preset fixtures validates golden plan structure for the Tampines MRT twins scenario 1`] = `
{
  "essentials": {
    "items": [
      {
        "helper": "Keeps you within walking distance and near childcare hubs.",
        "label": "Location radius",
        "value": "1.5km around Tampines MRT",
      },
      {
        "helper": "Set alerts at SGD 1.45M so turnkey options surface quickly.",
        "label": "Price band",
        "value": "SGD 1.2M â€“ 1.4M (stretch +5% for turnkey units)",
      },
      {
        "helper": "Filter for renovated family layouts with dual nursery potential.",
        "label": "Bedrooms & layout",
        "value": "3 bedrooms, turnkey finish",
      },
    ],
    "title": "Tune the essentials first",
  },
  "intent_summary": "Focused search: 3BR turnkey condo near Tampines MRT within ~SGD 1.4M.",
  "lifestyle_filters": {
    "helper": "Prioritise projects near preschools with playgrounds and service corridors.",
    "highlights": [
      "Childcare-ready developments",
      "Family-friendly facilities",
      "Low-renovation inventory",
    ],
    "title": "Lifestyle filters to toggle",
  },
  "micro_copy": {
    "follow_up": "Saved searches send weekly MRT-adjacent family picks plus preschool intel.",
    "reassurance": "Listings refresh daily; concierge flags new turnkey options instantly.",
  },
  "plan_actions": {
    "primary": {
      "description": "Opens a shortlist with MRT proximity and school filters applied.",
      "label": "Preview curated Tampines listings",
    },
    "secondary": [
      {
        "description": "Align financing buffers for turnkey units and twins arrival timeline.",
        "label": "Book PropertyGuru Finance consult",
      },
      {
        "description": "Compare childcare options and transport during a guided visit.",
        "label": "Schedule neighbourhood tour",
      },
    ],
  },
}
`;
